docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,198 @@
+* Can either modify KBP, or create a new tool/package to use instead of KBP
+
+1) Launch a management job called the *Job Manager* that sits in the KBP Queue, alongside other KBP jobs. Other jobs are launched in the NJS queue.
+2) Launch the *Setup Job* which will
+ * Use the *User Parameters* and/or
+ * Optionally Download the Data from the *User Parameters*  to figure out *Job Manager* parameters
+ * Use the results of information gathered from the initial download and or *User Parameters*
+ * Generate final parameters to be sent to the *Job Manager* to launch *Fan Out* jobs
+3) The *Job Manager* now has enough parameters to setup *Fan Out* Jobs, and will send out jobs and wait for them (and possibly retry them upon failure)
+4) *Fan Out* jobs download data and perform calculations, save them back to the system, and return references to the saved objects
+5) The *Job Manager* optionally launches a *Group/Reduce* job based on User Parameters and or the results of *Fan Out* Jobs
+6) The *Group/Reduce* job downloads objects from the system, and creates a set or other grouping, and then saves the set back to the system
+7) The *Job Manager*  sends the results back to a *Report* Job, which downloads the final data and uploads a report based on that data
+8) The *Job Manager* returns the reference to the results of the *Report Job*
+
+Pros/Cons
+* `-` Addresses the deadlocking issue, UI still broken for regular runs and batch runs if we re-use KBP
+* `-` Have to rewrite all apps that use KBP to use this new paradigm
+* `-` Will have to upload data and download data more times as no computation is done in the *Job Manager*
+* `+` All KBP jobs can run on a small subset of machines, deadlock issue is fixed
+* `+` No changes to ee2 required
+
+
+ 
+### Modify Apps to do only local submission by remove KBP, and moving the job 
+* `+` Simple solutions, quick turnarounds, fixes deadlock issue, fixes UI issues
+* `-` We have a limited number of larger resources machines
+* `-` Continued dependency on deprecated KBP tools
+* `-` App runs may take longer since fewer resources may be available to the app run
+
+### Deprecate kbparallels, and write new ee2 endpoints to entirely handle split and aggregate
+* `+` No longer uses an app, No longer uses a slot in the queue
+* `-` All jobs would have to change to stop using KBP in favor of these via the ee2 client,
+* `+` Can manage jobs more closely: have tables, more control over lifecycle, resource requests, canceling subjobs
+* `-` Job monitoring thread would have to run in ee2
+* `-`  Requires new data structures and models in ee2 to make sure that the relationship is preserved between jobs and sub jobs, and to make sure deadlocking does not occur
+* `-` Requires storing the job outside of condor to prevent job submission, unless we can mark jobs as runnable or not via the HTCondor API* 
+* `-` The endpoint would need to know how to split up input files and aggregate them
```

Comment:
I don't this this is true - the FO/FI jobs would do that, right?

Reply:
I thought this section was about https://github.com/kbase/execution_engine2/pull/422/files#diff-07a370a5473999326b83d1e3b2460686c97753837cfd37d1b78b12d809327ae5R63-R78

e.g. making EE2 do what the LKBP job would do, more or less, so you don't need the LKBP job (and probably makes UI stuff easier to handle if EE2 controls anything)

To me writing code in EE2 to figure out how to split up input for specific apps seems like a really bad idea; personally I wouldn't even have that as an option unless there's something I'm missing that makes it viable

Reply:
Was this comment not about line 198, in the alternatives considered section?

Reply:
Yeah, I didn't realize that we were considering (I guess?) 2 different EE2 based solutions, one where EE2 just manages the jobs much like the LKBP job would, and one where EE2 actually does the FI and FO parts itself. If that's really what's going on here to my mind we could just not include the latter and I'd be fine with it, seems pretty crazypants to me.

Again, maybe I don't get it though. Let's talk in the call tomorrow

Reply:
Ah, I think I see the ambiguity here. Basically instead of 
```
for item in data:
   params['job_input] = item
     ee2.run_job(params)
```
it would be
  ee2.run_fan_out_job(params, data)

```
def ee2_run_fan_out_job
    for item in data
    params[job_input] = item
   ee2.run_job()
```

Reply:
We decided to delete this section in an OOB meeting

Reply:
:+1: